lib/oli/delivery/sections.ex

```diff
@@ -2272,58 +2272,128 @@ defmodule Oli.Delivery.Sections do
     Repo.all(query)
   end
 
+  defp get_student_pages(section_slug, user_id, opts \\ [graded: false]) do
+    graded_filter =
+      case opts[:graded] do
+        true -> dynamic([_, _, _, _, rev], rev.graded == true)
+        _ -> []
+      end
+
+    SectionResource
+    |> join(:inner, [sr], s in Section, on: sr.section_id == s.id)
+    |> join(:inner, [sr, s], spp in SectionsProjectsPublications,
+      on: spp.section_id == s.id and spp.project_id == sr.project_id
+    )
+    |> join(:inner, [sr, _, spp], pr in PublishedResource,
+      on: pr.publication_id == spp.publication_id and pr.resource_id == sr.resource_id
+    )
+    |> join(:inner, [sr, _, _, pr], rev in Revision, on: rev.id == pr.revision_id)
+    |> join(:left, [sr], gc in GatingCondition,
+      on:
+        gc.section_id == sr.section_id and gc.resource_id == sr.resource_id and
+          is_nil(gc.user_id)
+    )
+    |> join(:left, [sr], gc2 in GatingCondition,
+      on:
+        gc2.section_id == sr.section_id and gc2.resource_id == sr.resource_id and
+          gc2.user_id == ^user_id
+    )
+    |> where(
+      [sr, s, _, _, _, gc, gc2],
+      s.slug == ^section_slug and (is_nil(gc) or gc.type == :schedule) and
+        (is_nil(gc2) or gc2.type == :schedule)
+    )
+    |> where(^graded_filter)
+    |> select([sr, s, _, _, rev, gc, gc2], %{
+      id: sr.id,
+      title: rev.title,
+      slug: rev.slug,
+      end_date:
+        fragment(
+          "cast(coalesce(coalesce(cast(? as text), cast(? as text)), cast(? as text)) as date) as end_date",
+          gc2.data["end_datetime"],
+          gc.data["end_datetime"],
+          sr.end_date
+        ),
+      scheduled_type: sr.scheduling_type,
+      gate_type:
+        fragment(
+          "coalesce(coalesce(cast(? as text), cast(? as text)), NULL) as hard_gate_type",
+          gc2.type,
+          gc.type
+        ),
+      graded: rev.graded,
+      resource_type_id: rev.resource_type_id,
+      resource_id: rev.resource_id,
+      numbering_level: sr.numbering_level,
+      children: sr.children,
+      section_id: s.id,
+      relates_to: rev.relates_to
+    })
+    |> order_by([{:asc_nulls_last, fragment("end_date")}])
+  end
+
+  @doc """
+    Returns the activities that a student need to complete next.
+  """
+  def get_next_activities_for_student(section_slug, user_id) do
+    get_student_pages(section_slug, user_id)
+    |> Repo.all()
```

Comment:
I think that it would make sense to move this step into the query that is executed in `get_student_pages`.  I think that it would involve adding a where clause comparing the end date against the current date and time, ordering by the date/time and then adding a `limit` of 2.   That will avoid this query (which has to run on every view of the course portal) from fetching all pages.